src/reconstruction/reconstructor.cpp

```diff
@@ -409,50 +422,100 @@ void reconstructor::initialize(acquisition::geometry geom) {
     }
 } // namespace slicerecon
 
-void reconstructor::transpose_sino_(std::vector<float>& projection_group,
-                                    std::vector<float>& sino_buffer,
-                                    int group_size) {
+/**
+ * In-memory CPU update of sino_buffer with group-buffered data, preparing for ASTRA
+ *
+ * @param projection_group  Reference to buffered data
+ * @param sino_buffer       Reference to the CPU buffer of the projections
+ * @param group_size        Number of projections in the group
+ */
+void reconstructor::transpose_sino_(float* projection_group,
+                                    float* sino_buffer,
+                                    int proj_offset,
+                                    int proj_end,
+                                    size_t buffer_size) {
     // major to minor: [i, j, k]
     // In projection_group we have: [projection_id, rows, cols ]
     // For sinogram we want: [rows, projection_id, cols]
 
     for (int i = 0; i < geom_.rows; ++i) {
-        for (int j = 0; j < group_size; ++j) {
+        for (int j = proj_offset; j <= proj_end; ++j) {
             for (int k = 0; k < geom_.cols; ++k) {
-                sino_buffer[i * group_size * geom_.cols + j * geom_.cols + k] =
+                sino_buffer[i * buffer_size * geom_.cols + j * geom_.cols + k] =
                     projection_group[j * geom_.cols * geom_.rows +
                                      i * geom_.cols + k];
             }
         }
     }
 }
 
-void reconstructor::upload_(int proj_id_min, int proj_id_max) {
+/**
+ * Process the projections [proj_id_start, ..., proj_id_end]
+ *
+ * @param proj_id_min
+ * @param proj_id_max
+ */
+void reconstructor::process_(int proj_id_begin, int proj_id_end) {
     if (!initialized_) {
         return;
     }
 
     slicerecon::util::log << LOG_FILE << slicerecon::util::lvl::info
-                          << "Uploading buffer (" << write_index_
-                          << ") between " << proj_id_min << "/" << proj_id_max
+                          << "Processing buffer (" << write_index_
+                          << ") between " << proj_id_begin << "/" << proj_id_end
                           << slicerecon::util::end_log;
 
+    auto data = buffer_[write_index_].data();
+
     environment_.spawn(parameters_.filter_cores, [&](auto& world) {
-        util::process_projection(
-            world, geom_.rows, geom_.cols, buffer_[write_index_].data(),
-            dark_.data(), flat_fielder_.data(), filter_, proj_id_min,
-            proj_id_max, !geom_.parallel, fdk_weights_,
-            !parameters_.already_linear, fft_plan_, ffti_plan_,
-            freq_buffer_[world.rank()], parameters_.retrieve_phase,
-            paganin_filter_, fft2d_plan_);
+        util::process_projection(world, geom_.rows, geom_.cols,
+                                 data, dark_.data(),
```

Comment:
Here we put `buffer_[write_index_]` data to be processed. This data only has size equal to `group_size` or `update_every`, yet in the process function you seem to index with 'global (GPU) projection indices' (i.e. not within the group).

Reply:
Good catch, that was exactly where the edit in `processing.cpp` was for :)